File: 297-serialize-and-deserialize-binary-tree/297-serialize-and-deserialize-binary-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public string serialize(TreeNode root) {
        StringBuilder sb = new();
        
        if (root == null)
            return sb.ToString();
        
        String leftSubtree = serialize(root.left);
        String rightSubtree = serialize(root.right);
        
        sb.Append($"({root.val},{leftSubtree.Length},{rightSubtree.Length})");
        sb.Append(leftSubtree);
        sb.Append(rightSubtree);
        
        return sb.ToString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(string data) {
        if (string.IsNullOrEmpty(data))
            return null;
        
        int index = data.IndexOf(')');
        string[] rootValues = data[1..index].Split(",");
        
        TreeNode root = new TreeNode(int.Parse(rootValues[0]));
        
        int leftStringLen = int.Parse(rootValues[1]);
        root.left = deserialize(data[(index + 1)..(index + 1 + leftStringLen)]);
        
        
        int rightStringLen = int.Parse(rootValues[2]);
        root.right = deserialize(data[(index + leftStringLen + 1)..(index + 1 + leftStringLen + rightStringLen)]);
        
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: 297-serialize-and-deserialize-binary-tree/297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        var sb = new StringBuilder();
        preorder(root, sb);
        sb.deleteCharAt(sb.length() - 1);
        System.out.println(sb.toString());
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] vals = data.split(",");
        return createTree(vals);
    }
    
    private void preorder(TreeNode root, StringBuilder sb) {
        if (root == null) {
            sb.append("N,");
            return;
        }
        
        sb.append(root.val);
        sb.append(',');
        
        preorder(root.left, sb);
        preorder(root.right, sb);
    }
    
    private int i = 0;
    
    private TreeNode createTree(String[] vals) {
        if (vals[i].equals("N")) {
            this.i++;
            return null;
        }
        
        var root = new TreeNode(Integer.parseInt(vals[i]));
        this.i++;
        
        root.left = createTree(vals);
        root.right = createTree(vals);
        
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));